# 0.2: The Falling Penny (Modeling with Computation)

*Modeling and Simulation in Python*

Copyright 2021 Allen Downey, (License: [Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International](https://creativecommons.org/licenses/by-nc-sa/4.0/))

Revised, Mike Augspurger (2021-present)

## The Falling Penny Myth

You might have heard that a penny dropped from the top of the Empire State Building would be going so fast when it hit the pavement that it would be embedded in the concrete; or if it hit a person, it would break their skull.

We can test this myth by making and analyzing a model. 

The model we'll use here is a pretty simply one: in fact, it will look a fair bit like a homework problem from a physics class.  This is an important connection to make, because indeed most of your homework problems in physics (or chemistry, or even biology) are models--that is, they are simplications of a real world situation designed to help us understand the real world situation better.

Since every model is a simplification, we start with some simplifying assumptions.  To start, we'll assume that the effect of air resistance is small. Because this effect is small, we will ignore it.  Now of course, this means are model will no longer be exactly correct, which might or might not be a problem, depending on how we hope to use the answer.  We can check later to see how "wrong" our model is.

OK. Now if air resistance is negligible, the primary force acting on the penny
is gravity, which causes the penny to accelerate downward.  We'll set the gravitational acceleration to $g = 9.8 m/s^2$.  Note that even *this* is a simplified model (!): $g$ actually varies with elevation, but again we're going to ignore that variation (this is a pretty good simplification--the error it creates is tiny).

If the initial velocity is 0 and the acceleration is $a = g$, the velocity after $t$ seconds is 

$$v = g t$$

and the distance the penny has dropped is 

$$x = g t^2 / 2$$ 

These two equations are known as *kinematic equations*, and you'll run into them in your physics classes, if you haven't already.

To find the time until the penny reaches the sidewalk, we can rearrange the second equation to solve for $t$:

$$t = \sqrt{ 2 x / g}$$ 

Plugging in a value for $g$ and the height of the Empire State Building ($x = 381 m$), we get $t = 8.8$ s. 

Then computing $v = g t$ we get a velocity on impact of $86$ m/s, which is about 190 miles per hour. That sounds like it could hurt.

So far, we've just done a pretty simple physics calcultion model: no need for a computer simulation yet.

But clearly this answer is not correct.  It turns out that ignoring air resistance is not a good choice. So we want to *iterate* our model: that is, we want to add in some new complication that will make the model more troublesome, but happily more accurate.  It turns out that once
the penny gets to about 29 m/s, the upward force of air resistance
equals the downward force of gravity, so the penny stops accelerating.
This is the *terminal velocity* of the penny in air.

Once the penny reaches terminal velocity, it doesn't matter how much farther it falls; it hits the sidewalk at about 29 m/s.
That's much less than 86 m/s, as the simple model predicts.

We'll use Python's computational tools in the rest of this chapter to simulate this more complex model. 

The statistician George Box famously said "All models are wrong, but
some are useful." He was talking about statistical models, but his wise words apply to all kinds of models. Our first model, which ignores air resistance, is very wrong, and probably not useful. The second model, which takes air resistance into account, is still wrong, but it's better, and it's good enough to refute the myth.

The television show *Mythbusters* has tested the myth of the falling
penny more carefully; you can view the results at <https://www.facebook.com/watch/?v=10152607367328224>. Their work provides a good sense of why this question is a lot more complicated than you might think.

## Computation In Python

Let me show you how I computed the results from the
previous section using Python.
First we'll create a variable to represent acceleration due to gravity in meters per second squared (m/s$^2$).

In [2]:
g = 9.8

A *variable* is a name that corresponds to a value. In this example, the name is `g` and the value is the number `9.8`.

Suppose we let the penny drop for $3.4$ seconds (s).  I'll create a variable to represent this time:

In [3]:

t = 3.4

Now we can compute the velocity of the penny after `t` seconds, and assign the answer to a variable named `v`:

In [4]:
v = g * t

Python uses the symbol `*` for multiplication.  The other arithmetic operators are `+` for addition, `-` for subtraction, `/` for division, and `**` for exponentiation.

When you assign a value to a variable, Jupyter does not show the result automatically, but you can display the value of a variable like this:

In [5]:
v

33.32

After $3.4$ s, the velocity of the penny is about $33$ m/s (ignoring air resistance).  Now let's see how far it would travel during that time.  Remember from above that our equation for distance travelled is $x = g t^2 / 2$:

In [6]:
x = g * t**2 / 2
x

56.644

It would travel about $56$ m.  Now, going in the other direction, let's compute the time it takes to fall 381 m, the height of the Empire State Building.

In [7]:
h = 381

For this computation, we need the square root function, `sqrt`, which is provided by a *library* called NumPy, which contains functions related to the manipulation of numbers.
Before we can use it, though,we have to *import* it.  Importing a function means we make it available to use, even though the code for that function is not in this particular notebook:

In [8]:
from numpy import sqrt

Again using the kinematic equation that we used above, we can write:

In [9]:
t = sqrt(2 * h / g)
t

8.817885349720552

With no air resistance, it would take about $8.8$ s for the penny to reach the sidewalk.

In [10]:
v = g * t
v

86.41527642726142

And its velocity on impact would be about $86$ m/s.

### Jupyter Tip

When you change the contents of a cell, you have to run it again for those changes to have an effect.  If you forget to do that, the results can be confusing, because the code you are looking at is not the code you ran.

If you ever lose track of which cells have run, and in what order, you should go to the Kernel menu and select "Restart & Run All".  Restarting the kernel means that all of your variables get deleted, and running all the cells means all of your code will run again, in the right order.

Select "Restart & Run All" now and confirm that it runs all of the cells.

### False Precision

Python displays results with about 16 digits, which gives the impression that they are very precise, but don't be fooled.
The numbers we get out are only as good as the numbers we put in.

For example, the "roof height" of the Empire State Building is $380$ m (according to Wikipedia: <https://en.wikipedia.org/wiki/Empire_State_Building>).
I chose $h=381$ m for this example on the assumption that the observation deck is on the roof and you drop the penny from a 1 meter railing.
But that's probably not right, so we should treat this value as an approximation where only the first two digits are likely to be right.

If the precision of the inputs is two digits, the precision of the outputs is two digits, *at most*.
That's why, if the output is `86.41527642726142`, I report it as "about 86".

### Units and Computation

The computations we just did use numbers without units.
When we set `h=381`, we left out the meters, and when we set `g=9.8`, we left out the meters per second squared.
And, when we got the result `v=86`, we added back the meters per second.  Leaving units of out computation is common practice, but it tends to cause errors, including the very expensive failure of the Mars Climate Orbiter in 1999 (see <https://en.wikipedia.org/wiki/Mars_Climate_Orbiter>).

We have three choices.  We could us a units library such as Pint (<https://pint.readthedocs.io/>).
This helps to find errors in the code because it recognizes unit mistakes, but I (and most programmers) find it unwieldy: it requires too much attention and mucks up the code.

Our second choice is to always use base units.  In the metric system, if I always put my values in $kg$ (kilogram),$m$ (meter), and $s$ (seconds), as well as other more complex base units like Newtons and Watts, I know that my answer will be in base units (presuming I set up my equations correctly).  This often requires using exponents, since a given simulation might include millions of seconds or a small fraction of a kg, but it allows us to generally avoid unit conversions. This is the approach we'll (mostly) use in this textbook.

Unfortunately, life does not always provide you with metric base units.  So in some cases, you need to convert data or solutions into more common units (like "miles per hour").  In those cases, you want to be extra careful in your conversions: always test your conversions by printing out a test case and making sure the code is converting correctly.

## Summary

As a first example, we modeled a penny dropped from the Empire State building, including gravity but ignoring air resistance.  In the exercises, you'll have a chance to try a better model, including air resistance.

### Exercise 1

In mathematical notation, we can write an equation like $v = g t$ and it's understood that we are multiplying $a$ and $t$.
But that doesn't work in Python.  If you put two variables side-by-side, like this:

```
v = g t
```

you'll get a *syntax error*, which means that something is wrong with the structure of the program.
Try it out in the next cell so you see what the error message looks like.

In [11]:
v = gt

NameError: name 'gt' is not defined

The error messages you get from Python can be big and scary (not this one), but if you read them carefully, they contain a lot of useful information.

The last line usually tells you what type of error happened, and sometimes additional information, so you might want to start from the bottom and read up.  Also look for the arrows (---->) in the error message.  Sometimes there are multiple arrows, but one of them will point to the line in the code that led to the error.

The previous lines are a *traceback* of what was happening when the error occurred.  The first section of the traceback shows the code you wrote.  The following sections are often from Python libraries.

### Exercise 2

In this chapter we used the `sqrt` function from the NumPy library.  NumPy also provides a variable named `pi` that contains an approximation of the mathematical constant $\pi$.
We can import it like this:

In [12]:
from numpy import pi
pi

3.141592653589793

NumPy provides other functions we'll use, including `log`, `exp`,`square`,`power`, `sin`, and `cos`.  To find out how to use these functions, see https://numpy.org/doc/stable/reference/routines.math.html

Import `sin` and `cos` from NumPy and compute

$$sin^2 (\pi/4) + cos^2 (\pi/4)$$

Note: A mathematical identity tells us that the answer should be $1$.  There are are multiple ways to do this, including using `power` or `square` from the NumPy library (remember to import them).  Try to find a couple ways to get the correct answer.  

In [13]:
from numpy import cos
from numpy import sin
from numpy import square, power

x = square(sin(pi/4))+ square(sin(pi/4))
y = power(sin(pi/4),2) + power(sin(pi/4),2)
z = (sin(pi/4))**2 + (sin(pi/4))**2

print(x,y,z)



1.0000000000000002 1.0000000000000002 1.0000000000000002


Did you notice that you often do not get exactly $1$?  This is the difference between *analytical math* (which tells us that the answer is exactly $1$) and *computational math* (which routinely has (generally) small rounding errors).


### Exercise 3

Suppose instead of dropping the penny, you throw it downward at its terminal velocity, $29$ m/s.  How long would it take to fall $381$ m?   (Hint: If we know a constant velocity and distance travelled, we can find the time.  Consider a similar question: if I have to drive 180 miles to Chicago, and drive 60 mph, how long will it take me?  How do I get that answer?)

In [14]:
# Complete the solution here.  The penny would fall at a constant velocity of 29 m/s.  

h = 381
v = 29
t = h/v
t

13.137931034482758

### Exercise 4:

So far we have considered two models of a falling penny:

* If we ignore air resistance, the penny falls with constant acceleration, and we can compute the time to reach the sidewalk and the velocity of the penny when it gets there.

* If we take air resistance into account, and drop the penny at its terminal velocity, it falls with constant velocity. 

Now let's consider a third model that includes elements of the first two: let's assume that the acceleration of the penny is `a` until the penny reaches $29$ m/s, and then $0$ m/s afterwards.  What is the total time for the penny to fall $381$ m?

This is a two phase simulation, because we have to break the problem into 2 time periods: when it is accelerating downward, and when its acceleration is zero.  To solve, answer the following 3 questions, using the kinematic equations at the beginning of the chapter:

1. How long would the penny take to reach $29$ m/s with constant acceleration `g`.  When an object is accelerating at $9.8  m/s^2$, that means every second its velocity increases by $9.8  m/s$.
2. How far would it fall during that time?
3. How long would it take to fall the remaining distance with constant velocity $29$ m/s?

Suggestion: Assign each intermediate result to a variable with a meaningful name. 

In [15]:
# Establish the unchanging variables (these are called "parameters")
g = 9.8
h = 381
v_term = 29

In [16]:
# Find how long it will take the penny to reach terminal velocity
t1 = v_term/g
t1

2.9591836734693877

In [17]:
# Find how far it will fall while it is reaching terminal velocity
h1 = (g*(t1**2))/2
h1

42.90816326530612

In [18]:
# Find how long it will take to fall the remaining distance at constant velocity
h2 = h - h1
t2 = h2/v_term
t2

11.658339197748065

In [19]:
# Determine the total time
t_total = t1 + t2
t_total

14.617522871217453

According to this model, it takes almost 15 seconds to fall to the ground, which is more than our first model (8.8 s) as well as our second model (13.1 s).

### Exercise 5

A pitcher on the baseball team claims that when he throws a fastball he is throwing the ball down; that is, the ball leaves his hand at a downward angle.  But watching from the side, an observor argues that the ball leaves his hand at an upward angle.

Can you think of a simple model you could use to settle the argument?  What factors would you include and what could you ignore?  What quantities would you have to look up or estimate?

I suggest you convert all quantities to SI units like meters and seconds (see <https://en.wikipedia.org/wiki/International_System_of_Units>).

In [20]:
# Start by naming your key parameters (the unchanging variables).  
# Do some estimating here (in British units).

# Release height based on pitcher's height
h_rel_B = 6     # feet

# Pitcher's mound height
h_mou_B = 10/12 # feet

# Catcher glove height
h_cat_B = 1     # feet

# Distance from release point to catcher's glove
d_thr_B = 60    # feet

# Initial velocity of pitch
v_hor_B = 80    # miles per hour

# Acceleration caused by gravity (metric)
g = 9.8         # m/s

In [21]:
# Convert your parameters to metric
# Test your velocity conversion: it should be 30-40 m/s, depending on the velocity of your pitch

foot_to_m = 0.3048
mile_to_m = 1609.34
hour_to_s = 3600


h_rel = h_rel_B * foot_to_m
h_cat = h_cat_B * foot_to_m
h_mou = h_mou_B * foot_to_m
d_thr = d_thr_B * foot_to_m
v_hor = v_hor_B * (mile_to_m/hour_to_s)
v_hor

35.76311111111111

In [22]:
# Assumption 1: no air resistance.  Find the time for the ball
# to reach the catcher's mitt

t = d_thr/v_hor
t

0.511364907353325

In [23]:
# Assumptions 2,3, and 4: Spin makes no difference, the ball starts out in 
# horizontal direction, and the ball falls vertically according to 
# gravitational acceleration.   Find how far the ball falls vertically 
# while in the air (assuming that its initial velocity in the vertical
# direction is zero)

d_fall = (g*(t**2))/2
d_fall


1.2813209355151258

In [24]:
# If the ball falls vertically more than the vertical distance between 
# release point and the mitt, then it must have been thrown at an 
# upward angle

# Find the vertical location of horizontally thrown ball when it reaches
# the catcher's mitt.  If this number is positive, an accurate pitch would need
# to be thrown at a downward angle (and vice versa)

d_hand_to_mitt = h_rel + h_mou - h_cat

loc_vertical_at_mitt = d_hand_to_mitt - d_fall
loc_vertical_at_mitt

0.4966790644848744

Your answer will vary according to your assumptions and parameter values.  I found that the horizontally-released pitch reached home plate about 0.5 meters above the catcher's mitt.
